driver-core/src/main/com/mongodb/internal/connection/SaslAuthenticator.java

```diff
@@ -55,7 +55,7 @@ public Void run() {
 
                     BsonInt32 conversationId = res.getInt32("conversationId");
 
-                    while (!(res.getBoolean("done")).getValue()) {
+                    while (!(res.getBoolean("done")).getValue() || !saslClient.isComplete()) {
```

Comment:
I don't think we should continue in the loop if the response says done.  Otherwise we're going to get a weird exception somewhere in here.  Instead, consider putting the check for isComplete() after the loop, and if it's false, through a security exception.

Reply:
Continuing despite the {done:1} is exactly what we want to do if the client hasn't completed the proof validation.  I do agree this is incomplete, however.  A server which sent an early done would (presumably) never send done again.  We should remember that done has been issued and stop when both that *and* isCompletely are true.

If I may suggest:

  ```suggestion
                        boolean done = res.getBoolean("done");
                        while (!done || !saslClient.isComplete()) {
                                response = saslClient.evaluateChallenge((res.getBinary("payload")).getData());
                                if (response == null) { ... yada yada ... }
                                
                                if (done && saslClient.isComplete()) break;
                                
                                res = sendSaslContinue(conversationId, response, connection);
                                done = done || res.getBoolean("done");
                          }                  
```

This will let us note the server's "doneness", but not until we feel done as well.

Reply:
I don't see why a client would send another response to a server that claims it's "done".  

Perhaps the client should check the final payload in a response which contains `done: true`, but I'm not sure if the protocol allows a payload in that case.

Reply:
All the examples in https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst where `done: true` contain empty payloads.  I can't find any other documentation.

Reply:
I have updated the code to use Sara's suggestion, and I have updated the async code along these same lines. I will wait for Sara to address Jeff's concerns with this change.

Reply:
Sara sent me the following reply:
It's not about sending a response so much as making sure the client has _locally_ processed the server's proof.   `evaluateChallenge()` doesn't send the reply, it processes what the server said and _prepares_ its next response to the server `sendSaslContinue()` is where the packet would be sent if the client and server aren't both done.

Reply:
I spoke with Jeff out of band, and I don't think that it's permissible for a well-behaved server to send {done: 1} multiple times, or even accept additional handshake messages from the client after it has done so. But, when a server *does* send {done: 1}, it may send a payload which the client needs to process in order for itself to complete.

SASL doesn't distinguish between a party being "done" with the conversation, and a party trusting its peer, I think this is a part of the layering separations between the mechanism and the application code running the conversation. If the server needs a message from the client to finish the conversation, then the server wasn't actually done. If the client sees {done: 1}, and doesn't have enough information to complete, it can't send further messages to the server because it will believe that the server has terminated the conversation and destroyed the AuthenticationSession with the mechanism state.

This appears to be covered by [RFC 4422](https://tools.ietf.org/html/rfc4422#section-3.6), which indicates that the concluding message must be sent by a server:
>    At the conclusion of the authentication exchange, the server sends a
>    message, the particulars of which are protocol specific, to the
>    client indicating the outcome of the exchange.

Reply:
Thanks Spencer.  Given that, I think we can simplify the algorithm a bit.

Reply:
I'm ready to resolve this thread, but will wait for Spencer and Sara to confirm.